/*
 * Decompiled with CFR 0.152.
 */
package jnr.posix;

import jnr.ffi.Pointer;
import jnr.ffi.Runtime;
import jnr.ffi.TypeAlias;
import jnr.posix.FreeBSDCmsgHdr;
import jnr.posix.SocketMacros;

public class FreeBSDSocketMacros
implements SocketMacros {
    public static final FreeBSDSocketMacros INSTANCE = new FreeBSDSocketMacros();

    @Override
    public int CMSG_LEN(int l) {
        return this.CMSG_ALIGN(FreeBSDCmsgHdr.layout.size()) + l;
    }

    public int CMSG_ALIGN(int len) {
        int sizeof_size_t = Runtime.getSystemRuntime().findType(TypeAlias.size_t).size();
        return len + sizeof_size_t - 1 & ~(sizeof_size_t + -1);
    }

    @Override
    public int CMSG_SPACE(int l) {
        return this.CMSG_ALIGN(FreeBSDCmsgHdr.layout.size()) + this.CMSG_ALIGN(l);
    }

    @Override
    public Pointer CMSG_DATA(Pointer cmsg) {
        return cmsg.slice(this.CMSG_ALIGN(FreeBSDCmsgHdr.layout.size()));
    }
}

